### PR TITLE
Split symbol interner into static unsynchronized and dynamic synchronized parts

### PR DESCRIPTION
**Pull request by sivadeilra**
Draft PR for investigation of improving Symbol lookups

I've been working on understanding parallelism in Rust, and I noticed one opportunity for improvement. That's in how we handle interning strings as Symbols. I decided to avoid placing all "static" symbols (symbols known at compile time) in the Interner table, and instead to just assign them static indices.

The advantage of this is that you can avoid taking a lock for resolution of static symbols. I measured, and for many Rust crates, the number of symbol lookups that can be resolved statically is between 45% and 75%. So, this could significantly reduce the amount of lock contention in Symbol::intern, if/when we ever get the compiler moved to parallel builds.

I've run ./x.py test src/tests/ui, and everything passes. But there's still some cleanup left to do, which is why I'm creating this as a draft PR.

### CODE REVIEW

**Comment by joshtriplett** [comment]
@Mark-Simulacrum CI isn't quite passing yet; I was going to queue it as soon as CI passed.

**Comment by panstromek** [comment]
Just to note, there was a PR in this area with similar goals but different approach (packing short strings into a pointer). It wasn't merged in the end, but I thought it's worth mentioning.
https://github.com/rust-lang/rust/pull/74554

**Comment by Mark-Simulacrum** [comment]
The try build failed (Option 'document-private-items' given more than once), FWIW, so perf won't run until that finishes successfully - the error is pretty odd, feels like maybe an implementation bug?

**Comment by petrochenkov** [comment]
#74554 wasn't merged because now we access the string representation of `Symbol`s rarely enough for it to not make much impact (at least in non-parallel case).

File: compiler/rustc_macros/src/symbols.rs [hunk 1]
```diff
@@ -1,12 +1,373 @@
...
+//! ```
+//!
+//! Unfortunately, this translates to a sequential scan. Our goal is to generate
+//! a function which is considerably more efficient, but produces the same result.
+//!
+//! Our strategy is divide and conquer: We want to quickly eliminate strings that
+//! are not static strings, and for the remaining candidates, to partition potential
+//! results into smaller sets as quickly as possible. We have the advantage of
+//! being able to compute, at (proc-macro) compile time, information about the
+//! candidate strings.
```
Comment (bjorn3) [inline compiler/rustc_macros/src/symbols.rs:180]: It would be nice if mir building could do this optimization. That would benefit everyone doing large matches on strings.

File: compiler/rustc_parse/benches/rustc_parse_bench.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+use rustc_session::parse::ParseSess;
+use rustc_span::{edition::Edition, SessionGlobals};
+use std::path::Path;
+
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+
+fn bench_parser(c: &mut Criterion) {
+    // moderately complex file
+    let input_file_path = Path::new(r"d:\rust\library\core\src\intrinsics.rs");
```
Comment (bjorn3) [inline compiler/rustc_parse/benches/rustc_parse_bench.rs:11]: This should be changed.

File: compiler/rustc_typeck/src/collect.rs [hunk 3]
```diff
@@ -860,6 +860,7 @@ fn convert_variant(
             let fid = tcx.hir().local_def_id(f.hir_id);
             let dup_span = seen_fields.get(&f.ident.normalize_to_macros_2_0()).cloned();
             if let Some(prev_span) = dup_span {
+                tracing::warn!("f = {}", f.ident.name.as_u32());
```
Comment (bjorn3) [inline compiler/rustc_typeck/src/collect.rs:863]:
```suggestion
```

**Comment by jyn514** [comment]
> The try build failed (Option 'document-private-items' given more than once), FWIW, so perf won't run until that finishes successfully - the error is pretty odd, feels like maybe an implementation bug?

cc https://github.com/rust-lang/rust/pull/73936

**Comment by petrochenkov** [comment]
@sivadeilra 
I took some liberty to rename the PR to what I think it actually does.
Feel free to rename it again to something better.

**Comment by petrochenkov** [comment]
First impressions:
- This will need to be split into, like, dozen of more focused PRs.
- Each part will need to go through perf, some of them probably wont be landed due to being on the wrong side of the benefits vs complexity trade-off.
- It would be great to use whole `crates.io` as the training set. Right now some of the static symbols look very domain-specific.
- I'm not sure the script doing the training run and producing the common symbol set belongs to rustc sources (especially if it's going to analyze crates.io).

**Comment by petrochenkov** [comment]
cc @nnethercote 
(I think you'll be interested in this.)

**Comment by petrochenkov** [comment]
>This will need to be split into, like, dozen of more focused PRs.

For initial PRs I'd suggest:
- Split interner into two parts for already existing groups of symbols (`kw` and `sym`).
- Add "common" symbols using statistics collected from crates.io (but not the script using for their collection).
- **No** divide and conquer (a simple `HashMap` instead), **no** single-byte optimization.

This way we get all optimizations associated with
- reduced locking
- somewhat reduced dynamic allocations (which shouldn't matter much because symbols use an arena with bump allocator).

**Comment by sivadeilra** [comment]
Ok, I've broken this work into several different PRs, some of which have already been committed. This PR now contains only the work to add the support for static symbols. It does not contain the "common words" optimization, or the single-character optimization. Those will be separate PRs, after this one.

Because parsing is _currently_ single-threaded, this PR has no effect on compiler throughput. But it does prepare us for eventually parallelizing parsing. Also, on a machine with many `rustc` processes running concurrently, those processes will be able to share the static symbol table, since it is mapped as read-only / shareable memory pages. That's better than having that portion of the symbol table duplicated in every process.

File: compiler/rustc_macros/src/symbols.rs [hunk 4]
```diff
@@ -151,51 +149,99 @@ fn symbols_with_errors(input: TokenStream) -> (TokenStream, Vec<syn::Error>) {
...
-        });
+        let symbol_index = if check_dup(keyword.name.span(), &value_string, &mut errors).is_ok() {
+            let symbol_index = symbol_strings.len() as u32;
+            symbol_strings.push(value_string);
+            symbol_index
+        } else {
+            // Compilation is going to fail, because of a duplicate symbol name.
+            // Add a symbol definition, even one with a bogus value, to minimize
+            // unrelated errors.
+            0
```
Comment (petrochenkov) [inline compiler/rustc_macros/src/symbols.rs:168]: If any value works here, then `symbol_index` would work too, so it can be moved out of the `check_dup` condition.

File: compiler/rustc_macros/src/symbols.rs [hunk 5]
```diff
@@ -151,51 +149,99 @@ fn symbols_with_errors(input: TokenStream) -> (TokenStream, Vec<syn::Error>) {
...
+        } else {
+            if check_dup(symbol.name.span(), &value, &mut errors).is_ok() {
+                let symbol_index = symbol_strings.len() as u32;
+                symbol_strings.push(value);
+                symbol_index
+            } else {
+                // Compilation is going to fail, because of a duplicate symbol name.
+                // Add a symbol definition, even one with a bogus value, to minimize
+                // unrelated errors.
+                0
```
Comment (petrochenkov) [inline compiler/rustc_macros/src/symbols.rs:220]: If any value works here, then `symbol_index` would work too, so it can be moved out of the `check_dup` condition.

File: compiler/rustc_span/src/symbol.rs [hunk 6]
```diff
@@ -1248,6 +1248,7 @@ symbols! {
         write_bytes,
         xmm_reg,
         ymm_reg,
+        zero_digit: "0",
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1251]: Not necessary in this PR.

File: compiler/rustc_macros/src/symbols.rs [hunk 7]
```diff
@@ -151,51 +149,99 @@ fn symbols_with_errors(input: TokenStream) -> (TokenStream, Vec<syn::Error>) {
...
+        check_order(symbol.name.span(), &name_string, &mut errors);
         let value = match &symbol.value {
             Some(value) => value.value(),
-            None => name.to_string(),
+            None => name_string,
+        };
+
+        let symbol_index = if let Some(i) = is_digit_symbol(&value) {
+            // We allow duplicates for digits, because we are not creating
+            // a new symbol index, just a new `sym::foo` name for an existing symbol.
```
Comment (petrochenkov) [inline compiler/rustc_macros/src/symbols.rs:209]: This exception shouldn't be necessary in this PR.

File: compiler/rustc_resolve/src/lib.rs [hunk 8]
```diff
@@ -2414,10 +2414,9 @@ impl<'a> Resolver<'a> {
                             (format!("maybe a missing crate `{}`?", ident), None)
                         }
                     } else if i == 0 {
-                        if ident
-                            .name
-                            .with(|n| n.chars().next().map_or(false, |c| c.is_ascii_uppercase()))
-                        {
+                        let ident_name_str = ident.name.as_str();
+                        let n = &*ident_name_str;
+                        if n.chars().next().map_or(false, |c| c.is_ascii_uppercase()) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/lib.rs:2419]:
```suggestion
                        if ident.name.as_str().chars().next().map_or(false, |c| c.is_ascii_uppercase()) {
```

File: compiler/rustc_span/src/symbol.rs [hunk 9]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
         Symbol(SymbolIndex::from_u32(n))
     }
 
+    /// Maps a string to its interned representation, but only if this string is a known
+    /// (static) symbol.
+    pub fn intern_static(string: &str) -> Option<Symbol> {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1452]:
```suggestion
    fn intern_static(string: &str) -> Option<Symbol> {
```

File: compiler/rustc_span/src/symbol.rs [hunk 10]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
...
+        if let Some(symbol) = STATIC_SYMBOLS_PHF.get(string) { Some(*symbol) } else { None }
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1453]:
```suggestion
        STATIC_SYMBOLS_PHF.get(string).copied()
```

File: compiler/rustc_span/src/symbol.rs [hunk 11]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
...
+    }
+
     /// Maps a string to its interned representation.
+    #[inline(never)]
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1457]: Why is this necessary?

File: compiler/rustc_span/src/symbol.rs [hunk 12]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
...
-    /// requires locking the symbol interner.
-    pub fn with<F: FnOnce(&str) -> R, R>(self, f: F) -> R {
-        with_interner(|interner| f(interner.get(self)))
+    pub fn is_static(self) -> bool {
+        self.0.as_u32() < DYNAMIC_SYMBOL_BASE
+    }
+
+    /// Translates the `Symbol` to a string, but only if this `Symbol`
+    /// was originally interned as a static symbol.
+    pub fn as_str_static(self) -> Option<&'static str> {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1472]:
```suggestion
    fn as_str_static(self) -> Option<&'static str> {
```

File: compiler/rustc_span/src/symbol.rs [hunk 13]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
...
+        let symbol_index = self.0.as_usize();
+        if symbol_index < STATIC_SYMBOLS.len() {
+            // This is a well-known symbol. The symbol string is stored in a static field.
+            // There is no need to lock the interner.
+            Some(STATIC_SYMBOLS[symbol_index])
+        } else {
+            None
+        }
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1480]:
```suggestion
        STATIC_SYMBOLS.get(self.0.as_usize()).copied()
```

File: compiler/rustc_span/src/symbol.rs [hunk 14]
```diff
@@ -1446,23 +1447,54 @@ impl Symbol {
...
-        with_interner(|interner| unsafe {
-            SymbolStr { string: std::mem::transmute::<&str, &str>(interner.get(self)) }
-        })
+        if let Some(string) = self.as_str_static() {
+            SymbolStr { string }
+        } else {
+            // This is a dynamic string. The string is stored in the Interner.
+            with_interner(|interner| unsafe {
+                let string = interner.get_dynamic(self);
+                SymbolStr { string: std::mem::transmute::<&str, &str>(string) }
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1495]:
```suggestion
                SymbolStr { string: std::mem::transmute::<&str, &str>(interner.get_dynamic(self)) }
```

File: compiler/rustc_span/src/symbol.rs [hunk 15]
```diff
@@ -1484,19 +1516,22 @@ impl Symbol {
 
 impl fmt::Debug for Symbol {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.with(|str| fmt::Debug::fmt(&str, f))
+        let s: &str = &*self.as_str();
+        fmt::Debug::fmt(s, f)
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1520]:
```suggestion
        fmt::Debug::fmt(&self.as_str(), f)
```

File: compiler/rustc_span/src/symbol.rs [hunk 16]
```diff
@@ -1484,19 +1516,22 @@ impl Symbol {
...
     }
 }
 
 impl fmt::Display for Symbol {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.with(|str| fmt::Display::fmt(&str, f))
+        let s: &str = &*self.as_str();
+        fmt::Display::fmt(s, f)
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1527]:
```suggestion
        fmt::Display::fmt(&self.as_str(), f)
```

File: compiler/rustc_span/src/symbol.rs [hunk 17]
```diff
@@ -1484,19 +1516,22 @@ impl Symbol {
...
     }
 }
 
 impl<S: Encoder> Encodable<S> for Symbol {
     fn encode(&self, s: &mut S) -> Result<(), S::Error> {
-        self.with(|string| s.emit_str(string))
+        let string: &str = &*self.as_str();
+        s.emit_str(string)
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1534]:
```suggestion
        s.emit_str(&self.as_str())
```

File: compiler/rustc_span/src/symbol.rs [hunk 18]
```diff
@@ -1536,21 +1578,32 @@ pub struct Interner {
 }
 
 impl Interner {
-    fn prefill(init: &[&'static str]) -> Self {
-        Interner {
-            strings: init.into(),
-            names: init.iter().copied().zip((0..).map(Symbol::new)).collect(),
-            ..Default::default()
-        }
+    pub fn fresh() -> Self {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:1581]: `Interner::default` already exists, so this method can be removed.

**Comment by petrochenkov** [comment]
The CI error ("tidy error: Dependencies not explicitly permitted") means that `phf` and the other new compiler dependencies need to be added to the whitelist in `src\tools\tidy\src\deps.rs` (assuming the licenses are compatible).

File: compiler/rustc_span/src/symbol.rs (hunk 18 above)
Comment (sivadeilra) [reply compiler/rustc_span/src/symbol.rs:1581]: Removed.

**Comment by sivadeilra** [comment]
> The CI error ("tidy error: Dependencies not explicitly permitted") means that `phf` and the other new compiler dependencies need to be added to the whitelist in `src\tools\tidy\src\deps.rs` (assuming the licenses are compatible).

Thanks. I checked the license for the affected crates. All are available under MIT and/or Apache 2.0 licenses. I've added them to the allow-list.

File: compiler/rustc_span/src/symbol.rs (hunk 6 above)
Comment (sivadeilra) [reply compiler/rustc_span/src/symbol.rs:1251]: Removed.

File: compiler/rustc_macros/src/symbols.rs (hunk 5 above)
Comment (sivadeilra) [reply compiler/rustc_macros/src/symbols.rs:220]: Done.

File: compiler/rustc_macros/src/symbols.rs (hunk 4 above)
Comment (sivadeilra) [reply compiler/rustc_macros/src/symbols.rs:168]: Done.

**Comment by sivadeilra** [comment]
@petrochenkov: A quick question on PR etiquette: For the issues above, where I have made modifications in response to feedback, should I click "Resolve conversation"? Or is that something that the reviewer generally does, after reviewing the modifications made in response to PR feedback? I ask because different communities generally have different policies, and I'm not familiar with how the Rust community uses this tool.

**Comment by petrochenkov** [comment]
@sivadeilra 
I don't think we have an explicit policy about this, some people resolve conversations some don't.
For me as a reviewer it's more convenient if they are not collapsed, because I'm double checking everything anyway.

**Comment by petrochenkov** [comment]
Some comments were hidden by GitHub and not addressed.

I'd start perf run now if one of them didn't involve `#[inline(never)]` which may affect the results.
Looks like in the single-threaded case here we are trading a possibly faster string lookup from PHF for a poorly predictable branch, I want to make sure that the perf diff is at least within the noise.

**Comment by sivadeilra** [comment]
About the `#[inline(never)]`. I've found that excessive inlining can lead to code bloat, without any performance benefit. In fact, in some cases, inlining leads to worse performance, because the same code is duplicated, and because it can needlessly complicate control flow (and on register-poor machines like x86/x64 to lead to more register spilling) at the call sites.

Functions like `String::intern` are good candidates for suppressing inlining, for several reasons: 1) They're called many times, so the cost of the duplication can become significant. 2) They are not generic, so inlining has fewer benefits, 3) The function itself does more "internal" work than most small functions, when you consider the code underneath it that gets inlined into it (such as the `phf` lookup and the `FxHashMap` lookup).

I did my own local perf runs, and there is no difference in perf between the versions with and without `#[inline(never)]`. However, the version with `#[inline(never)]` is noticeably smaller (about 2 KB). I know that doesn't seem like much, but excessive inlining can be a non-linear source of bloat, so it's something I'm always watching for.

**Comment by petrochenkov** [comment]
Could you add a comment to `fn intern` explaining why the `#[inline(never)]` was used?

**Comment by sivadeilra** [comment]
> Could you add a comment to `fn intern` explaining why the `#[inline(never)]` was used?

Sure.

**Comment by sivadeilra** [comment]
Added comment, rebased on origin/master.

File: compiler/rustc_macros/src/symbols.rs (hunk 1 above)
Comment (sivadeilra) [reply compiler/rustc_macros/src/symbols.rs]: Yes! I've written some prototype code for handling `match` over `&str` with lots of string-literal patterns. It would be great if `rustc` did this work, and then the special-case code in `symbols.rs` could be eliminated entirely.

If I can produce something reasonable, I'll work up a separate PR for that.

**Comment by bjorn3** [comment]
Big regressions of up to 3.6%.

**Comment by sivadeilra** [comment]
> Big regressions of up to 3.6%.

That's distressing, of course. I'd like to understand exactly what is being measured, since my own local measurements showed this as being perf neutral (or I wouldn't have bothered to submit the PR, of course). Obviously I'm not testing the same thing.

What's the best way to examine one of these cases, locally, so that I can do profiling runs that more closely match what the lab is testing?

**Comment by lqd** [comment]
You can run the perf collector locally https://github.com/rust-lang/rustc-perf/blob/master/collector/README.md

**Comment by bjorn3** [comment]
It is possible that changes on master caused this PR to become a regression. You should try to benchmark the build for https://github.com/rust-lang/rust/commit/eaab320bc5d8187ab57f3a0b91133a9608ed2ef5 which the merge commit that was benchmarked by `@rust-timer`. You can use [rustup-toolchain-install-master](https://github.com/kennytm/rustup-toolchain-install-master) to install it.

**Comment by petrochenkov** [comment]
Top regressions being from rustdoc likely means that `Symbol::intern` got slower (maybe due to two string searches instead of one).

(Explanation: reading metadata from other crates is a large source of `Symbol::intern`s, metadata is read on demand, and rustdoc happens to demand almost all of it.)

**Comment by petrochenkov** [comment]
I'll try to do some experiments with single-threaded performance on our standard benchmark in https://github.com/rust-lang/rust/pull/80420.

**Comment by petrochenkov** [comment]
Ok, I think I know how to mitigate the regressions.

- Use `FxHashMap<&'static str, Symbol>` for both static and dynamic tables.
- Use [`HashMap::raw_entry`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.raw_entry) to avoid hashing the string twice (example: 88091185ef4a4df837c3c607686d4b28703b9d61).
- Create the static table dynamically, but "before `main`" using [`ctor`](https://crates.io/crates/ctor) so it doesn't require synchronization.

Comments:
- Basic interning algorithm for reference:
	```rust
	if let Some(symbol) = static_table.get(string) {
		symbol
	} else {
		dynamic_table.get(string)
	}
	```
- Our standard benchmark works with a single-threaded compiler on Linux, together with [initial-exec TLS model](https://github.com/rust-lang/rust/pull/78201) used by the compiler it means that access to the dynamic table is cheap, so we can assume that we are basically just searching two hash tables consequently (with some large probability, e.g. 30-50%), this is unsurprisingly slower than searching one table.
- How to make these two searches faster? Find a common part in them and perform it once. If these two tables use a common hash table type (including a common hasher), then the common part is string hashing. The Raw Entry API allows to perform hashing only once (example: 88091185ef4a4df837c3c607686d4b28703b9d61).
- Besides that `phf::Map` happens to be slower than regular `FxHashMap` for search (https://github.com/rust-lang/rust/pull/80420#issuecomment-752718785), so it would be faster to use `FxHashMap` even without deduplicated hashing.
- `ctor` is pretty un-Rusty, but we are using it with an eye on replacing it with compile-time initialization when `const_heap` (https://github.com/rust-lang/rust/issues/79597) and `const_trait_impl` (https://github.com/rust-lang/rust/issues/67792) work well enough. (If this was C++, using a global variable with a constructor would be a no-brainer here.)
- The results with this strategy are good enough compared to other variants (https://github.com/rust-lang/rust/pull/80420#issuecomment-753385047), with couple of outliers like `deeply-nested-async-*` or `clap-rs-doc`. I didn't investigate why the outliers happen.

**Comment by JohnCSimon** [comment]
@sivadeilra Ping from trigage can you please address the merge conflict and post the status of this PR?

**Comment by sivadeilra** [comment]
> @sivadeilra Ping from triage can you please address the merge conflict and post the status of this PR?

Sure, will do.

**Comment by sivadeilra** [comment]
I think this design direction has merit, but it currently has one cost that is not acceptable. For single-threaded execution, it does penalize symbol lookups, because for many symbols it will query two hash tables, not just one. This cost could be reduced somewhat, but it's an inherent part of the design, so it can't be eliminated.

Since one of the big sources of symbol lookups is importing metadata from dependent crates, I've been planning on doing an experiment, where we serialize symbols differently depending on whether they are "static" (known) or dynamic. For static symbols, we can simply write the symbol index, and we don't even need to write the string. During deserialization, we would simply use the symbol index, without any need to check any interning tables.  I think it's possible that this would have a measurable performance benefit, and that the benefit might outweigh the cost of the double-lookup.

@JohnCSimon, would you prefer that I abandon this PR, or change it to "draft", while I work on that experiment? It seems that this PR cannot be accepted in its current form, which I think is fine.  Either that, or I could use `#[cfg(feature = "static_symbols")]` to gate this feature. That way, it could be committed (and so the PR wouldn't bit-rot while other changes go in), without any cost to others, while I work on this design.

**Comment by Dylan-DPC** [comment]
@sivadeilra i would recommend closing this, working on the experiment and based on that you can reopen or create a new PR with these changes. Since the experiment will take some it would ensure that the PR doesnt stay open and rot with conflicts

**Comment by sivadeilra** [comment]
@Dylan-DPC Agreed. Closing this now.
